src/main/java/software/amazon/smithy/lsp/ext/SmithyProject.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.lsp.ext;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.eclipse.lsp4j.Location;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import software.amazon.smithy.lsp.SmithyInterface;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.SourceLocation;
+import software.amazon.smithy.model.shapes.ShapeType;
+import software.amazon.smithy.model.validation.ValidatedResult;
+
+public final class SmithyProject {
+    private List<Path> imports;
+    private List<File> smithyFiles;
+    private List<File> externalJars;
+    private Map<String, List<Location>> locations = Collections.emptyMap();
+    private ValidatedResult<Model> model;
+    private File root;
+
+    private SmithyProject(List<Path> imports, List<File> smithyFiles, List<File> externalJars, File root,
+                          ValidatedResult<Model> model) {
+        this.imports = imports;
+        this.root = root;
+        this.model = model;
+        this.smithyFiles = smithyFiles;
+        this.externalJars = externalJars;
+        model.getResult().ifPresent(m -> {
+            this.locations = collectLocations(m);
+        });
+    }
+
+    /**
+     * Recompile the model, optionally adding a file to the tracked list of sources.
+     *
+     * @param file file which may or may not be already tracked by this project
+     * @return either an error, or a loaded project
+     */
+    public Either<Exception, SmithyProject> recompile(File file) {
+        // We aggressively re-build the model with only existing files
```

Comment:
I'm happy with it, but I'd rather it was mutating the current SmithyProject but created a new one.

Reply:
> I'd rather it was mutating the current SmithyProject but created a new one
I'm struggling to parse this.

Do you want

a) mutate in-place (model, locations, smithyFiles)
b) build a new one each time?

Reply:
Sorry I meant "I'd rather it was not mutating", so yeah I'd like the `recompile` method to **not** mutate in place.

Reply:
Ok, cool, that's what it's doing at the moment, only overhead is compilation + locations re-creation, the coursier re-fetching doesn't happen.

In the future we can look into updating existing values rather than rebuilding from scratch, if it becomes a problem.

Reply:
hah, I completely misread the code, my bad